### PR TITLE
Add curl to send json

### PR DESCRIPTION
**Pull request by vitsalis**

### CODE REVIEW

File: bot_prototype.php [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<?php
+    echo $_POST[ 0 ];
+?>
```
Comment (vitsalis) [inline bot_prototype.php]: In the spec you said there would be a bot prototype as an example. I suggest we put it here.

Comment (dionyziz) [inline bot_prototype.php]: No, let's create a new folder "bots". In it, let's have a folder for every language, one should be "php". In it we will have the example and library for PHP.

File: models/game.php [hunk 2]
```diff
@@ -183,6 +185,23 @@ public function jsonSerialize() {
             }
             return compact( 'gameid', 'W', 'H', 'M', 'MAX_HP', 'players' );
         }
+
+
+        public function sendJson() {
```
Comment (dionyziz) [inline models/game.php]: Networking logic should not be in game model. Make a different class responsible for communicating with bots. Game model should be agnostic about methods of communication.

File: models/round.php [hunk 3]
```diff
@@ -60,6 +60,22 @@ public function jsonSerialize() {
             return compact( 'round', 'map' );
         }
 
+        public function sendJson() {
```
Comment (dionyziz) [inline models/round.php]: Similar.

File: models/test/withfixtures.php [hunk 4]
```diff
@@ -17,5 +20,31 @@ protected function buildCountry( $name, $shortname ) {
 
             return $country;
         }
+        protected function buildRound() {
```
Comment (dionyziz) [inline models/test/withfixtures.php:57]: Good that we're keeping fixtures separated here.

File: bots/php/.htaccess [hunk 5]
```diff
@@ -0,0 +1,2 @@
+RewriteEngine On
+RewriteRule bot bot.php
```
Comment (dionyziz) [inline bots/php/.htaccess:2]: Needs rewrite rules for round, game.

File: bots/php/round.php [hunk 6]
```diff
@@ -0,0 +1,4 @@
+<?php
```
Comment (dionyziz) [inline bots/php/round.php:1]: Needs game.php

File: bots/php/round.php [hunk 7]
```diff
@@ -0,0 +1,4 @@
...
+    $map = $_POST[ 'map' ];
+    echo json_encode( $intents );
```
Comment (dionyziz) [inline bots/php/round.php:3]: `$intents` is not defined.

File: controllers/game.php [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+            $game->save();
+            $users = User::findAll();
+            $grader = new Grader( $users, $game );
+            $grader->initiate();
+            $grader->createGame();
+        }
+        public function createView() {
+            include 'views/game/create.php';
+        }
+        public function update( $grader ) {
```
Comment (dionyziz) [inline controllers/game.php]: How do you expect to pass an object to the controller?

File: controllers/game.php [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+            $grader->nextRound();
+            $grader->game->nextRound();
```
Comment (dionyziz) [inline controllers/game.php:24]: Shouldn't this be called from the grader?

File: database/migration/20140212210823_add_errors_table.php [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    include 'migrate.php';
+
+    migrate(
+        [
+            'CREATE TABLE IF NOT EXISTS
+                `errors` (
+                    `gameid` int(11) NOT NULL,
+                    `userid` int(11) NOT NULL,
+                    `error` varchar(20) COLLATE utf8_unicode_ci NOT NULL,
+                    CONSTRAINT pk_errors PRIMARY KEY (gameid,userid,error)
```
Comment (dionyziz) [inline database/migration/20140212210823_add_errors_table.php:11]: Avoid PK on varchars. Create an `errorid` autoincrement instead?

File: database/migration/20140212210823_add_errors_table.php [hunk 11]
```diff
@@ -0,0 +1,16 @@
+<?php
```
Comment (dionyziz) [inline database/migration/20140212210823_add_errors_table.php:1]: Update schema.

File: models/curl.php [hunk 12]
```diff
@@ -0,0 +1,41 @@
+<?php
+    interface CurlConnectionInterface {
+        public function __construct();
+        public function setOpt( $option, $value );
+        public function exec();
+        public function __destruct();
+    }
+
+    class CurlConnection implements CurlConnectionInterface {
```
Comment (dionyziz) [inline models/curl.php:9]: basic unit test

File: models/grader/bot.php [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+            }
+            $this->version = $decodedResponse->version;
+            $this->botname = $decodedResponse->botname;
+        }
+        public function sendGameRequest( $game ) {
+            try {
+                $ch = $this->httpRequest( 'game', 'create', GraderSerializer::gameRequestParams( $game ) );
+            }
+            catch ( CurlException $e ) {
+                throw new GraderBotException( end( $this->errors ) );
```
Comment (dionyziz) [inline models/grader/bot.php]: You should append the error to `$this->errors`

File: models/grader/grader.php [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    include_once 'models/grader/error.php';
+
+    class Grader {
+        public $registeredBots;
+        public $registeredUsers;
+        protected $game;
+        protected $users;
+        public $bots = [];
+
+        public function __construct( $users, $game, $graderBotObject = 'GraderBot' ) {
```
Comment (dionyziz) [inline models/grader/grader.php]: Rename `$graderBotObject` to `$graderBotClass`.

File: models/grader/error.php [hunk 15]
```diff
@@ -0,0 +1,28 @@
+<?php
+    class Error extends ActiveRecordBase {
+        protected static $attributes = [ 'gameid', 'userid', 'error' ];
+        protected static $tableName = 'errors';
+        protected $gameid;
+        protected $userid;
+        public $error;
+
+        public function __construct( $gameid, $userid, $error ) {
```
Comment (dionyziz) [inline models/grader/error.php]: Construction of an Error instance with arguments should look it up in the database; to construct an active record object associated with a new database record to-be-created using `save()`, use `new Error` and set attributes on the new object, as we do with other classes.

File: tests/models/grader/bot.php [hunk 16]
```diff
@@ -0,0 +1,447 @@
...
+            $this->assertFalse( $result[ 'caught' ], 'A GraderBotException must not be caught if bot responds with valid json' );
+            $this->assertTrue( empty( $result[ 'errors' ] ), 'There should be no errors if the json is valid' );
+        }
+        public function testInitiateRespondInvalidJson() {
+            $result = $this->initiateWithJsonAndGetErrors( '{ invalid_json }' );
+
+            $this->assertTrue( $result[ 'caught' ], 'A GraderBotExcpetion must be caught when response has invalid json' );
+            $this->assertEquals( 'initiate_invalid_json', $result[ 'errors' ][ 0 ], 'Bot who has invalid json as a response must have a "initiate_invalid_json" error' );
+        }
+        protected function initiateWithJsonAndGetErrors( $json ) {
```
Comment (dionyziz) [inline tests/models/grader/bot.php:158]: Move this function definition above its calls?

File: models/grader/bot.php [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+            foreach ( $requiredAttributes as $attribute ) {
+                if ( !isset( $decodedResponse->$attribute ) ) {
+                    $this->errors[] = 'round_' . $attribute . '_not_set';
+                    throw new GraderBotException( end( $this->errors ) );
+                }
+            }
+            if ( count( ( array )$decodedResponse ) > count( $requiredAttributes ) ) {
+                $this->errors[] = 'round_additional_data';
+                throw new GraderBotException( end( $this->errors ) );
+            }
```
Comment (dionyziz) [inline models/grader/bot.php:171]: So what do you do with the bot's response? Shouldn't you be putting the intents on the creatures somewhere? Or at least recording the intents here and putting them on the creatures in the caller?

File: tests/models/grader/grader.php [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+            $this->assertEquals( count( $grader->registeredBots ), count( $grader->registeredUsers ), 'The number of registered users must be the same with the number of registered bots' );
+
+            $error1 = Error::findErrorsByGameAndUser( $game->id, $users[ 0 ]->id );
+            $error2 = Error::findErrorsByGameAndUser( $game->id, $users[ 1 ]->id );
+
+            $this->assertTrue( empty( $error1 ), 'A user with valid boturl must not have errors' );
+            $this->assertTrue( empty( $error2 ), 'A user with valid boturl must not have errors' );
+        }
+        public function testCreateGameInvalidResponse() {
+            $game = new Game(); $game->save();
```
Comment (dionyziz) [inline tests/models/grader/grader.php]: nit: new line between statements

File: tests/models/grader/grader.php [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+            $this->assertTrue( !empty( $error2 ), 'A user with invalid game response must have errors' );
+
+            $this->assertEquals( 1, count( $error1 ), 'A user must get one error if they have an invalid game response' );
+            $this->assertEquals( 1, count( $error2 ), 'A user must get one error if they have an invalid game response' );
+
+            $this->assertEquals( 'game_error', $error1[ 0 ]->error, 'A user must get a "game_error" error if he has an invalid game response' );
+            $this->assertEquals( 'game_error', $error2[ 0 ]->error, 'A user must get a "game_error" error if he has an invalid game response' );
+        }
+        public function testCreateGameValidResponse() {
+            $game = new Game(); $game->save();
```
Comment (dionyziz) [inline tests/models/grader/grader.php]: nit: new line between statements

File: tests/models/grader/grader.php [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+            $grader->createGame();
+
+            $error1 = Error::findErrorsByGameAndUser( $game->id, $users[ 0 ]->id );
+            $error2 = Error::findErrorsByGameAndUser( $game->id, $users[ 1 ]->id );
+
+            $this->assertTrue( empty( $error1 ), 'A user with valid game response must not have errors' );
+            $this->assertTrue( empty( $error2 ), 'A user with valid game response must not have errors' );
+        }
+        public function testNextRoundInvalidResponse() {
+            $game = new Game(); $game->save();
```
Comment (dionyziz) [inline tests/models/grader/grader.php]: nit: new line between statements

**Comment by dionyziz** [comment]
Please add some integration tests for Grader and Game/resolution. Make a test that builds a game, adds some bots with appropriate mocks, goes through genesis and initialization, and then requests a round from the bots, then it asserts that the map has evolved in the next round based on the bots' responses.

File: database/migration/20140212210823_add_errors_table.php (hunk 10 above)
Comment (vitsalis) [inline database/migration/20140212210823_add_errors_table.php:11]: why should an error have an id?

Comment (dionyziz) [inline database/migration/20140212210823_add_errors_table.php:11]: Because we need a primary key.

File: tests/models/grader/bot.php [hunk 21]
```diff
@@ -0,0 +1,610 @@
...
+
+            $caught = false;
+            try {
+                $bot->sendInitiateRequest();
+            }
+            catch ( GraderBotException $e ) {
+                $caught = true;
+            }
+
+            $this->assertFalse( $caught, 'A GraderBotException should not be thrown if the username is correct' );
```
Comment (dionyziz) [inline tests/models/grader/bot.php]: DRY: This function has two parts that do almost the same thing. Make it into a protected function.

File: tests/models/grader/bot.php [hunk 22]
```diff
@@ -0,0 +1,610 @@
...
+        public function testInitiateRespondAdditionalData() {
+            $result = $this->initiateWithJsonAndGetErrors( json_encode( [
+                'botname' => 'suprabot',
+                'version' => '0.1.0',
+                'username' => 'vitsalis',
+                'additional' => 'shit'
+            ] ) );
+
+            $this->assertTrue( $result[ 'caught' ], 'A GraderBotExcpetion must be caught when response has additional data' );
+            $this->assertEquals( 'initiate_additional_data', $result[ 'errors' ][ 0 ], 'Bot whose username is not set must have a "initiate_additional_data" error' );
```
Comment (dionyziz) [inline tests/models/grader/bot.php]: This function body is very typical. Shorten and DRY:

```
$this->assertInitiationThrows( [
   'botname' => 'suprabot',
   'version' => '0.1.0',
   'username' => 'vitsalis',
   'additional' => 'shit'
], 'initiate_additional_data' );
```

and define `assertInitiationThrows` as a protected function in this class. Similarly for round and game.

File: tests/models/grader/grader.php [hunk 23]
```diff
@@ -0,0 +1,246 @@
...
+            $this->assertTrue( empty( $errors[ 'error1' ] ), 'A user with valid round response must not have errors' );
+            $this->assertTrue( empty( $errors[ 'error2' ] ), 'A user with valid round response must not have errors' );
+        }
+        public function testRoundResolution() {
+            $game = new Game();
+            $game->users = $this->users;
+            $game->initiateAttributes();
+            $game->id = 1;
+            $game->exists = true;
+            $creature1 = $this->buildCreature( 0, 1, 1, $this->users[ 1 ], $game );
```
Comment (dionyziz) [inline tests/models/grader/grader.php:189]: Creatures should be addressed from 1.

File: tests/models/grader/grader.php [hunk 24]
```diff
@@ -0,0 +1,246 @@
...
+            $creature2 = $this->buildCreature( 1, 2, 2, $this->users[ 2 ], $game );
+            $round = new Round();
+            $round->creatures = [ $creature1, $creature2 ];
```
Comment (dionyziz) [inline tests/models/grader/grader.php:192]: Key should be creatureid.

File: tests/models/resolution.php [hunk 25]
```diff
@@ -49,6 +49,7 @@ public function testMoveSingleCreature() {
                         $this->assertEquals( $creature->locationx - 1, $newCreature->locationx, 'A creature must move left by one when moving west' );
                         $this->assertEquals( $creature->locationy, $newCreature->locationy, 'A creature must not move to the y-direction when moving west' );
                         break;
+            $game->initiateAttributes();
```
Comment (dionyziz) [inline tests/models/resolution.php:52]: What's this?

File: tests/models/round.php [hunk 26]
```diff
@@ -1,42 +1,34 @@
...
+        public function testSaveDb() {
+            $game = $this->buildGame();
+            $game->rounds[ 0 ] = $round = $this->buildRound();
+            $round->game = $game;
+            $round->save();
+            $dbRound = new Round( $game, 1 );
 
-            $json = $round->toJson();
-            $data = json_decode( $json );
+            $this->assertEquals( $round->id, intval( $dbRound->id ), "Round's id must be correctly stored in the database" );
```
Comment (dionyziz) [inline tests/models/round.php:13]: `assertSame`

File: tests/models/round.php [hunk 27]
```diff
@@ -1,42 +1,34 @@
...
+            $this->assertEquals( $round->game->id, intval( $dbRound->game->id ), "Round's gameid must be correctly stored in the database" );
```
Comment (dionyziz) [inline tests/models/round.php:14]: `assertSame`

**Pull request merged**